### PR TITLE
`Discussion`: Add button to show / hide post replies

### PR DESCRIPTION
**Pull request by lschlesinger**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

User feedback lead to the change request that the previous UI for toggling the answers of a post is not intuitive / learnable.
With this new UI we strive for better understanding and consistency between posts without answers (show `Start Discussion` button) and with answers (did not show any button).

### Description

There will always be a button below the post content. 
**We distinguish following use cases:**
1. There is no answer yet: `Start Discussion` Button visible in post footer, clicking it will open the modal
2. There are answers, but not expanded yet: `Show X Reply` Button visible with answer count, clicking it will expand the answers (turning to state described in 3.)
3. There are answers, they are expanded: `Reply Now` Button visible, clicking it will open the modal, _AND_ `Hide Answers` Button visible, clicking it will hide the answers (turning back to state described in 2.)

### Steps for Testing

1. Log in to Artemis
2. Navigate course discussion overview / lecture page or exercise page
3. Find a post to reenact the use cases described in the **Description** section
4. Write a new post and check if in the similar posts shown when providing a post title, there are no buttons.

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

Old UI for posts without, with 1, with 2 answers:

https://user-images.githubusercontent.com/41366522/139492154-669ad2bb-bafb-4dd9-a5db-ee42ba5ffd86.mov

New UI for posts without, with 1, with 2 answers:

https://user-images.githubusercontent.com/41366522/139491398-e406163a-836c-472e-b635-9606d5842e8a.mov

### CODE REVIEW

File: src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts [hunk 1]
```diff
@@ -120,36 +121,58 @@ describe('PostingsThreadComponent', () => {
...
+    it('answer now button should not be visible answer posts are not shown', () => {
+        component.post = post;
+        component.post.answers = unsortedAnswerArray;
+        component.showAnswers = false;
+        fixture.detectChanges();
+        const answerNowButton = fixture.debugElement.nativeElement.querySelector('button');
+        expect(answerNowButton.innerHTML).not.toContain('answerNow');
+    });
+
+    it('answer now button should be visible answer posts are not shown', () => {
```
Comment (julian-christl) [inline src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts:161]:
```suggestion
    it('answer now button should be visible answer posts are shown', () => {
```

File: src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts [hunk 2]
```diff
@@ -120,36 +121,58 @@ describe('PostingsThreadComponent', () => {
...
+        expect(answerNowButton.innerHTML).toContain('answerNow');
     });
 
     it('should contain a post', () => {
         component.post = post;
         component.post.answers = unsortedAnswerArray;
         component.ngOnInit();
         const postComponent = fixture.debugElement.nativeElement.querySelector('jhi-post');
-        expect(postComponent).toBeTruthy();
+        expect(postComponent).not.toBe(undefined);
```
Comment (julian-christl) [inline src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts:175]:
```suggestion
        expect(postComponent).not.toBe(null);
```

File: src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts [hunk 3]
```diff
@@ -158,7 +181,7 @@ describe('PostingsThreadComponent', () => {
         component.showAnswers = true;
         fixture.detectChanges();
         const answerPostComponent = fixture.debugElement.nativeElement.querySelector('jhi-answer-post');
-        expect(answerPostComponent).toBeTruthy();
+        expect(answerPostComponent).not.toBe(undefined);
```
Comment (julian-christl) [inline src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts:184]:
```suggestion
        expect(answerPostComponent).not.toBe(null);
```

**Review by julian-christl — CHANGES_REQUESTED** [review]
Nice, love the changes.

I have some smaller remarks regarding the code.

File: src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html [hunk 4]
```diff
@@ -13,6 +13,19 @@
                         (sortedAnswerPosts.length === 0 ? 'artemisApp.metis.startDiscussion' : 'artemisApp.metis.answerNow') | artemisTranslate
                     }}</span>
                 </button>
+                <!-- expand answers -->
+                <button *ngIf="!showAnswers && sortedAnswerPosts.length > 0" class="btn btn-sm btn-outline-secondary mt-2" (click)="showAnswers = true">
+                    <fa-icon [icon]="'comments'"></fa-icon>
+                    <span class="d-none d-md-inline">
+                        {{ 'artemisApp.metis.showAnswers' | artemisTranslate: { number: sortedAnswerPosts.length } }}
+                        {{ (sortedAnswerPosts.length === 1 ? 'artemisApp.metis.answerSingular' : 'artemisApp.metis.answerPlural') | artemisTranslate }}</span
```
Comment (Olimbia23) [inline src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html:21]:
```suggestion
                        {{ 'artemisApp.metis.showAnswer' + (sortedAnswerPosts.length === 1 ? 'singular' : 'plural') | artemisTranslate: { number: sortedAnswerPosts.length} }}</span
```

**Review by Olimbia23 — COMMENTED** [review]
Code changes look good.
Nice work @lschlesinger!
I left one small suggestion regarding the translation.

**Comment by imstreetspirit** [comment]
tested on ts2, functionality works as expected. I like this version 👍 I can re-test after you address the code review suggestions

File: src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts (hunk 2 above)
Comment (lschlesinger) [reply src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts:175]: 👍 my mistake

File: src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts (hunk 1 above)
Comment (lschlesinger) [reply src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts:161]: good catch

File: src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts [hunk 5]
```diff
@@ -120,36 +121,58 @@ describe('PostingsThreadComponent', () => {
...
-        component.toggleAnswers();
-        expect(component.showAnswers).toEqual(true);
+        fixture.detectChanges();
+        const answerNowButton = fixture.debugElement.nativeElement.querySelector('button');
+        expect(answerNowButton.innerHTML).toContain('showAnswers');
+        expect(answerNowButton.innerHTML).toContain('answerPlural');
     });
 
-    it('answer now button should not be visible if post does not yet have any answers', () => {
+    it('should display button to show answer', () => {
```
[the remark this answers is not in the document]
Comment (lschlesinger) [reply src/test/javascript/spec/component/shared/metis/postings-thread/postings-thread.component.spec.ts:134]: 👍

File: src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html (hunk 4 above)
[the remark this answers is not in the document]
Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html]: changed it

Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html]: I changed the logic here, it's easier to read now

File: src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html [hunk 6]
```diff
@@ -13,6 +13,19 @@
                         (sortedAnswerPosts.length === 0 ? 'artemisApp.metis.startDiscussion' : 'artemisApp.metis.answerNow') | artemisTranslate
```
[the remark this answers is not in the document]
Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html:20]: done

**Comment by lschlesinger** [comment]
@julian-christl @Olimbia23 thanks for the feedback, I could integrate all of it.

**Review by Olimbia23 — DISMISSED** [review]
Nice work @lschlesinger!
Thank you for incorporating the changes.
Code looks good.
Approve 👍

File: src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html [hunk 7]
```diff
@@ -6,8 +6,15 @@
                 <ng-container *ngIf="showAnswers">
                     <jhi-answer-post *ngFor="let answerPost of sortedAnswerPosts" [posting]="answerPost"></jhi-answer-post>
                 </ng-container>
-                <!-- new answer -->
-                <button *ngIf="showAnswers || sortedAnswerPosts.length === 0" class="btn btn-sm btn-outline-secondary mt-2" (click)="createAnswerPostModal.open()">
+                <!-- start discussion -->
+                <button *ngIf="sortedAnswerPosts.length === 0" class="btn btn-sm btn-outline-secondary mt-2" (click)="createAnswerPostModal.open()">
+                    <fa-icon [icon]="'comments'"></fa-icon>
+                    <span class="d-none d-md-inline">{{
+                        (sortedAnswerPosts.length === 0 ? 'artemisApp.metis.startDiscussion' : 'artemisApp.metis.answerNow') | artemisTranslate
```
Comment (julian-christl) [inline src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html:13]:
```suggestion
                        'artemisApp.metis.startDiscussion' | artemisTranslate
```

File: src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html [hunk 8]
```diff
@@ -6,8 +6,15 @@
...
+                    }}</span>
+                </button>
+                <!-- add new answer to expanded discussion -->
+                <button *ngIf="showAnswers && sortedAnswerPosts.length > 0" class="btn btn-sm btn-outline-secondary mt-2" (click)="createAnswerPostModal.open()">
                     <fa-icon [icon]="'comments'"></fa-icon>
                     <span class="d-none d-md-inline">{{
                         (sortedAnswerPosts.length === 0 ? 'artemisApp.metis.startDiscussion' : 'artemisApp.metis.answerNow') | artemisTranslate
```
Comment (julian-christl) [inline src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html:20]:
```suggestion
                        'artemisApp.metis.answerNow' | artemisTranslate
```

File: src/main/webapp/i18n/de/metis.json [hunk 9]
```diff
@@ -7,10 +7,9 @@
             "addNewPost": "Neuer Beitrag",
             "answerNow": "Jetzt beantworten",
             "startDiscussion": "Diskussion beginnen",
-            "showAnswers": "Zeige {{ number }}",
+            "showMultipleAnswers": "Zeige {{ number }} Antworten",
+            "showSingleAnswer": "Zeige die Antwort",
```
Comment (julian-christl) [inline src/main/webapp/i18n/de/metis.json:11]:
```suggestion
            "showSingleAnswer": "Zeige 1 Antwort",
```
I'd suggest keeping the style of the one above for two reasons:
1. To stay consistent
2. It's not clear what "die" refers to. "Die" is quite direct. It should rather be "eine" and in that case, we can stick to the format

Comment (julian-christl) [reply src/main/webapp/i18n/de/metis.json:11]:
```suggestion
            "showSingleAnswer": "Zeige Antwort",
```
Alternatively, we can follow along with the English translation.
It doesn't bother me what but either choose one or the other (and change the other translation respectively to stay consistent.

**Review by julian-christl — COMMENTED** [review]
You forgot to resolve the conditionals and I have two remarks regarding the new translations.

File: src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html (hunk 7 above)
Comment (lschlesinger) [reply src/main/webapp/app/shared/metis/postings-thread/postings-thread.component.html]: yep...thanks

File: src/main/webapp/i18n/de/metis.json (hunk 9 above)
Comment (lschlesinger) [reply src/main/webapp/i18n/de/metis.json]: Changed it to "Zeige Antwort", that's better, you're right

**Review by julian-christl — APPROVED** [review]
Code looks good!

**Review by Olimbia23 — APPROVED** [review]
Reapprove 👍

**Review by imstreetspirit — APPROVED** [review]
tested on ts2 again, still works as expected.

**Pull request merged**